Patient registry search

lomamech opened the pull request:
Patient Registration Researche in a Modal
- Creation of a new component that display the filtered Elements
- Adding clean property in the util service for cleaning the data from the client side
- Removing the property required and bh-date-datepicker-editor to template of bh-date-editor
- Update components bhDateEditor and adding new property in function for the modale using
- Complete translate file

@jniles

lomamech committed:
Resolve Conflict with branch Master

lomamech committed:
Complete translate file and improve patient registry controller

lomamech committed:
Update conditon of expect in test E2E for billing service and debtor group

lomamech committed:
Complete translate file

jniles commented on client/src/js/components/bhFiltersApplied.js:
```diff
@@ -0,0 +1,14 @@
+angular.module('bhima.components')
+.component('bhFiltersApplied', {
+  controller: FiltersAppliedController,
```
I do not think that you need this controller, since it doesn't do anything.  What happens if you remove it?

jniles commented on client/src/js/services/util.js:
```diff
@@ -105,4 +105,18 @@ function UtilService($filter) {
   service.minimumDate = new Date('1900-01-01');
+
+  // utility function
+  service.clean = function clean(o) {
```
I think @sfount has an easier way of doing this.  @sfount, is this the case?

jniles commented on client/src/partials/patients/registry/modal.js:
```diff
@@ -0,0 +1,100 @@
+function PatientRegistryModalController( $uibModalInstance, Inventory, patients, util) {
+  var vm = this;
+
+  vm.period = {
```
This looks like a very useful utility!  In the future, we should put this in a service so that many controllers can benefit from it.

jniles commented on client/src/partials/patients/registry/modal.js:
```diff
@@ -0,0 +1,100 @@
+    year : {
+      cacheKey : 'year',
+      translateKey : 'FORM.BUTTONS.THIS_YEAR',
+      id : 'year'
```
Do you need an `id` with each of these records?  Where do you use it?

lomamech replied:
It used during test E2E
eg. element(by.id('year')).click();

jniles replied:
Thank you for clarifying.  From [this line](https://github.com/IMA-WorldHealth/bhima-2.X/pull/403/files#diff-7affaeca06dfed7a1ae3d4f1d6df6c00R22) I see that the HTML element id is set to the `cacheKey` property, instead of the `id` property.  Is it used somewhere else that I cannot find?

jniles commented on client/src/partials/patients/registry/modal.js:
```diff
@@ -0,0 +1,100 @@
+    });    
+  }
+
+  function setTimes(times){
```
I think these methods already exist in the `DateService`. They would look like this:

``` js

PatientRegistryModalController.$inject = [
  /* other injects ..., */ 'DateService'
];

function PatientRegistryModalController(/* injects ,*/ Dates) {

  // controller code ..

  // set the duration of the query
  function setTimes(times) {

   // the dateRegistrationTo never changes, so we set it at the beginning
   vm.patient = {
     dateRegistrationTo : new Date()
   };

   switch (times) {
      case 'today' :
        vm.patient.dateRegistrationFrom = new Date();
        break;
      case 'week' :
        vm.patient.dateRegistrationFrom = Dates.previous.week();
        break;
      case 'month' :
        vm.patient.dateRegistrationFrom = Dates.previous.month();
        break;
      default:
        vm.patient.dateRegistrationFrom = Dates.previous.year();
  }
}
```

jniles commented on client/src/partials/patients/registry/registry.js:
```diff
@@ -49,6 +51,29 @@ function PatientRegistryController($translate, Patients) {
+    });       
+  }
+
+  function momentAge(patientAge){
```
Nice refactor.

jniles commented on client/src/partials/patients/registry/registry.js:
```diff
@@ -49,6 +51,29 @@ function PatientRegistryController($translate, Patients) {
   }
 
+  // Search and filter data in Patiens Registry
+  function search() {
```
Could you put the modal in the `PatientService`?  That way, it can be called from multiple pages in necessary.  Something like:

``` js
// Patients is the PatientService
function search() {
    Patients.openSearchModal()
    .then(function (data) {
      // etc..
    });
}
```

jniles commented on client/src/partials/templates/bhDateEditor.tmpl.html:
```diff
@@ -31,11 +31,10 @@
         data-date-editor-input
         ng-readonly="!$ctrl.editMode"
         ng-disabled="$ctrl.disabled"
-        required>
```
This change will break some pages!  Could you use `ng-required` (see [documentation](https://docs.angularjs.org/api/ng/directive/ngRequired)) instead, and have it required by default?  That way only controllers do not need required can turn it off.

jniles commented on client/test/e2e/debtors/debtorgroups.spec.js:
```diff
@@ -11,7 +11,7 @@ var FU = require('../shared/FormUtils');
   'use strict';
 
-  let initialGroups = 3;
+  let initialGroups = 7;
```
Are you rebuilding your database each time?  I think you need to... this should always be 3.

jniles commented on client/test/e2e/billingServices/billingServices.spec.js:
```diff
@@ -84,6 +84,6 @@ describe('Billing Services', function () {
       .all(by.repeater('(rowRenderIndex, row) in rowContainer.renderedRows track by $index'));
 
-    expect(rows.count()).to.eventually.equal(2);
+    expect(rows.count()).to.eventually.equal(3);
```
This change breaks the end to end tests.

jniles commented on client/test/e2e/patient/registry.spec.js:
```diff
@@ -2,20 +2,118 @@
     expect(rows.count()).to.eventually.be.equal(defaultVisibleRowNumber);
   });
 
+  it('Search patient with the Name The Grid should be have 1 visible rows', function () {
```
Nice Tests!

jniles commented on client/test/e2e/shared/components/bhDateEditor.js:
```diff
@@ -15,8 +15,8 @@ module.exports = {
    */
-  set: function set(date, id) {
-
+  set: function set(date, id, elementClick) {
```
Could you update the documentation comments above this function?  It is missing a third `@param` declaration.

jniles commented on client/test/e2e/shared/FormUtils.js:
```diff
@@ -64,6 +64,8 @@ ElementArrayFinder.prototype.enabled = function () {
 var buttons =  {
   create : function create() { return $('[data-method="create"]').click(); },
+  create : function create() { return $('[data-method="create"]').click(); },
+  search : function search() { return $('[data-method="search"]').click(); },
```
This is a nice addition!

jniles commented on server/controllers/medical/patients/index.js:
```diff
@@ -392,61 +399,114 @@ function search(req, res, next) {
+    } 
+
+    if(qDob) {
+      // Research from patient registration dates
```
I think this is research from date of birth, not registration dates.

lomamech committed:
Applying recommandation

Delete unused controller in bhFiltersApplied
Adding new function in DateService
Adding and improving Patient Service
Restaure test E2E for billing Service and Debtor Groups
Add comment in components bhDateEditor
Upfdate and Improve comment in API patient in server side

jniles commented on client/src/partials/patients/registry/modal.js:
```diff
@@ -0,0 +1,66 @@
+        case 'month' :
+          vm.patient.dateRegistrationFrom = dateService.previous.month();
+          break;
+        default:
```
Very nice. :)
